Introduce tablet load balancer

tgrabiec opened the pull request:
After this series, tablet replication can handle the scenario of bootstrapping new nodes. The ownership is distributed indirectly by the means of a load-balancer which moves tablets around in the background. See docs/dev/topology-over-raft.md for details.

The implementation is by no means meant to be perfect, especially in terms of performance, and will be improved incrementally.

The load balancer will be also kicked by schema changes, so that allocation/deallocation done during table creation/drop will be rebalanced. 

Tablet data is streamed using existing `range_streamer`, which is the infrastructure for "the old streaming". This will be later replaced by sstable transfer once integration of tablets with compaction groups is finished. Also, cleanup is not wired yet, also blocked by compaction group integration.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2364/

avikivity commented:
spelling: "Intoroduce"

avikivity commented on test/boost/token_metadata_test.cc:
```diff
@@ -44,31 +44,6 @@ namespace {
-        BOOST_REQUIRE_EQUAL(erm->get_endpoints_for_reading(t1), inet_address_vector_replica_set{e1});
-    }
-}
-
```
@gusev-p you created this test (8877641b0f0cee2791a321cf0928a857ccccfbcc), please review this patch.

avikivity replied:
@bhalevy please review it as well.

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -116,6 +116,9 @@ public:
+    explicit operator bool() const {
+        return !empty();
+    }
     size_t size() const {
```
Disagree, chunked_vector is a replacement for std::vector and shouldn't extend its API.

tgrabiec replied:
Ok.

avikivity commented on locator/tablets.hh:
```diff
@@ -55,15 +55,33 @@ struct tablet_info {
+    allow_write_both_read_old,
+    write_both_read_old,
+    write_both_read_new,
+    streaming,
```
Shouldn't write_both_read_new be after streaming?

tgrabiec replied:
Yes. Not that it matters for correctness, since transition order is defined elsewhere.

avikivity replied:
Still, it helps reduce confusion.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -263,9 +263,29 @@ class tablet_effective_replication_map : public effective_replication_map {
+            case tablet_transition_stage::allow_write_both_read_old:
+            case tablet_transition_stage::write_both_read_old:
+            case tablet_transition_stage::write_both_read_new:
+            case tablet_transition_stage::streaming: {
```
That's out of scope of this series, because there is a lot of preexisting code like this.

avikivity replied:
https://github.com/scylladb/scylladb/pull/14607

tgrabiec replied:
I'll add [[fallthrough]] here.

avikivity commented on locator/tablets.cc:
```diff
@@ -263,9 +263,29 @@ class tablet_effective_replication_map : public effective_replication_map {
+                return to_replica_set(replicas);
+            }
+        }
+        abort();
```
These should be on_internal_error()s.

tgrabiec replied:
Even though the compiler guarantees that we can't reach this place (there is no default case)?

avikivity replied:
The compiler only guarantees we don't forget an enum. We can still have enums with bad values.

Also, it's a way to document that this case as considered (on_internal_error vs. on_fatal_internal_error).

avikivity commented on locator/tablets.cc:
```diff
@@ -165,6 +165,10 @@ std::ostream& operator<<(std::ostream& out, tablet_id id) {
     return out << size_t(id);
 }
 
+std::ostream& operator<<(std::ostream& out, global_tablet_id id) {
```
Please have the changelog explain what a global tablet_id is. In what scope? node or cluster? Why is it needed?

tgrabiec replied:
There is a comment above global_tablet_id class which explains the scope.

avikivity replied:
Ok. The changelog should mention it too.

avikivity commented on locator/tablets.cc:
```diff
@@ -165,6 +165,10 @@ std::ostream& operator<<(std::ostream& out, tablet_id id) {
 
+std::ostream& operator<<(std::ostream& out, global_tablet_id id) {
+    return out << id.table << ":" << id.tablet;
+}
```
Please use formatters in new code.

avikivity commented on message/messaging_service.hh:
```diff
@@ -182,7 +182,8 @@ enum class messaging_verb : int32_t {
     RAFT_TOPOLOGY_CMD = 64,
     RAFT_PULL_TOPOLOGY_SNAPSHOT = 65,
-    LAST = 66,
+    RAFT_STREAM_TABLET = 66,
```
Not related to raft except incidentally. Suggest to open a  new prefix TABLET_ for tablet operations.

tgrabiec replied:
Will change to TABLET_STREAM_DATA

avikivity commented on service/storage_service.cc:
```diff
@@ -5348,6 +5349,95 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+                _tablet_streaming.erase(tablet_);
+                auto ep = std::current_exception();
+                slogger.warn("Streaming for tablet migration of {} from {} failed: {}", tablet_, leaving_replica_, ep);
+                std::rethrow_exception(std::move(ep));
```
"throw;" works just as well, and then you don't need "ep".

tgrabiec replied:
Right, I just copied it from somewhere

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,405 @@
+        ///
+        /// The algorithm produces a migration plan which is a set of instructions about which tablets to move
+        /// where. The plan is a small increment, not a complete plan. To achieve balance, the algorithm should
+        /// be invoked iteratively until an empty plan is returned.
```
Does the plan have to be fully executed before we start another one? Or can we keep generating plans until we run out of things to do (given current transitions)?

avikivity replied:
Ack

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,405 @@
+                }
+                tid = *tmap.next_tablet(tid);
+            }
+        }
```
We should keep the structure updated based on events, but this is good enough to start.

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,405 @@
+                        }
+                    }
+                }
+                tid = *tmap.next_tablet(tid);
```
How do you know this isn't nullopt?

avikivity commented on replica/tablet_mutation_builder.hh:
```diff
@@ -0,0 +1,45 @@
+
+    mutation build() {
+        return std::move(_m);
+    }
```
Why can't we use CQL to update the tablets table?

avikivity replied:
@kostja let's work to improve that instead of building mutations by hand.

avikivity commented:
Looks good. I didn't review deeply the algorithm because (a) it's very long and complicated (b) we're going to replace it anyway with the edge generation thing.

tgrabiec commented:
CI fails due to https://github.com/scylladb/scylladb/issues/14614

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2377/

tgrabiec commented:
> CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2377/

tablets_test caught a scenario where load balancing oscillates and doesn't converge.

avikivity commented:
> > CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2377/
> 
> tablets_test caught a scenario where load balancing oscillates and doesn't converge.

Well, it should have some hysteresis.

tgrabiec commented:
In this case, oscillation is just an omission in logic. After fixing it, we stop with `unable to achieve balance`. But then the problem is that in this case it should be possible to achieve balance if the algorithm was smarter. The problem is with rack load constraints. If you have an overloaded node in one rack and target underloaded node in a different rack, we won't be able to reduce the load gap by moving tablets between the two. We have to balance the overloaded rack first, which is unconstrained.

avikivity commented:
> In this case, oscillation is just an omission in logic. After fixing it, we stop with `unable to achieve balance`. But then the problem is that in this case it should be possible to achieve balance if the algorithm was smarter. The problem is with rack load constraints. If you have an overloaded node in one rack and target underloaded node in a different rack, we won't be able to reduce the load gap by moving tablets between the two. We have to balance the overloaded rack first, which is unconstrained.

The algorithm can be as silly as we like for the first iteration. IMO, it's more important to have a good interface for the algorithm to drive (e.g. generate edges, including the ability to cancel if conditions change). The algorithm itself will be self-contained and easy to iterate on.

tgrabiec commented:
In V2:
  - use fmt::formatter<> instead of operator<<()
  - Dropped chunked_vector::operator bool() patch
  - Fixed load-balancer to not enter into oscillations by adding this hunk:
    ``` 
            // Prevent load inversion which can lead to oscillations.
            if (src_node_info.get_balanced_max_load(nodes[src_host].tablet_count - 1) <
                    nodes[target].get_balanced_max_load(nodes[target].tablet_count + 1)) {
                lblogger.debug("No more candidate nodes");
                break;
            }```

 - Changed RPC name to TABLET_STREAM_DATA
 - Reworked stream_tablet() to use the main coroutine flow and RAII instead of with_gate(...)
 - Simplified std::rethrow_exception(ep) into `throw;`
 - Fixed confusing order of enum values of tablet_transition_stage and switch cases 
 - Added `[[fallthrough]];`
 - Replaced abort() with on_internal_error()
 - Improved commit log message which introduces global_tablet_id
 - Used explicit type in `topo.for_each_node([&] (const locator::node* node_ptr) {` 
 - Documented guarantee that migration_info is unique per tablet in each migration_plan, on which the executing state machine relies.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2408/

tgrabiec commented:
In V3:
  - make load balancer more precise by using true average load instead of div_ceil() and make it more aggressive by trying to equalize it rather than allowing a difference of 1 tablet per shard. This fixes a regression in test_load_balancing_with_two_empty_nodes where load would not get balanced because the algorithm decides that's it's enough. Even though we may not need this level of precision, I want to keep at least one of the tests which verifies that the load is equally balanced.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2431/

tgrabiec commented:
CI blocked by https://github.com/scylladb/scylladb/issues/14675 and https://github.com/scylladb/scylladb/issues/14614

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2453/

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5377,6 +5378,96 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+    auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, tm, _abort_source,
+           get_broadcast_address(), _sys_ks.local().local_dc_rack(),
+           "Tablet migration", streaming::stream_reason::tablet, std::move(tables));
+    streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
```
Is it correct to stream from any replica of this tablet?

I think we should be following similar logic as for vnode replica sets, in order to preserve read-your-write consistency, i.e.:
- when we decommission a node, if that node was a replica for some tablet, we must stream from this replica and not any other
- but if some replica is dead and we remove it, we can stream from any replica -- or better, perform repair.

The requirement being that if we perform a quorum write followed by a quorum read, the read should observe the write. (In general if write_cl + read_cl > num_replicas)

tgrabiec replied:
We don't stream from any replica, we stream from the leaving replica:
```
    ranges_per_endpoint[leaving_replica_ip].emplace_back(range);
```

tgrabiec replied:
Handling of removenode is not implemented yet.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5377,6 +5378,96 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+    ranges_per_endpoint[leaving_replica_ip].emplace_back(range);
+    streamer->add_rx_ranges(table.schema()->ks_name(), std::move(ranges_per_endpoint));
+
+    if (_tablet_streaming[tablet]) {
```
All that work with creating `streamer` and so on looks unnecessary if we enter this branch -- we're waiting for an existing `streamer` to finish in this case, IIUC. Should the branch be placed earlier in the function?

tgrabiec replied:
It can be placed earlier. Not that it matters, since this is rarely called. It was easier that way with the futurize_invoke() version, but now it doesn't matter so I can move it.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5102,6 +5102,7 @@ void storage_service::on_update_tablet_metadata() {
     }
```
>  storage_service: Notify topology state machine after applying schema change

commit message: s/may allocator/may allocate/

kbr-scylla reviewed (changes requested):
@tgrabiec I suggest splitting the PR into two:
- `Introduce tablet load balancer` and commits that depend on it stay here,
- all other commits get extracted to a separate preparatory PR

This should make the review process a bit more efficient, and we should be able to merge the preparatory PR earlier (is it also blocked by the CI issues?), then focus only on reviewing the load balancer itself.

tgrabiec commented on locator/abstract_replication_strategy.cc:
```diff
@@ -112,10 +112,10 @@ inet_address_vector_topology_change vnode_effective_replication_map::get_pending
     return endpoints;
```
It currently doesn't support it though, because `vnode_effective_replication_map::get_natural_endpoints()` throws if token metadata has no normal tokens. So it's incorrect to create erm with such token metadata.

tgrabiec commented on locator/tablets.cc:
```diff
@@ -135,6 +135,32 @@ const tablet_transition_info* tablet_map::get_tablet_transition_info(tablet_id i
+});
+
+sstring tablet_transition_stage_to_string(tablet_transition_stage stage) {
+    return tablet_transition_stage_to_name.at(stage);
```
Ok

tgrabiec commented on mutation_writer/multishard_writer.cc:
```diff
@@ -43,6 +43,7 @@ class shard_writer {
 class multishard_writer {
 private:
     schema_ptr _s;
+    const dht::sharder& _sharder;
```
The enclosing scope of multishard_writer ensures liveness. 

This is a legacy from pre-tablet era, where schema::get_sharder() returns sharder&. We could change it, but it seems out of scope here.

avikivity replied:
I'd like to solve this in a different way. If we don't read past across tablets boundaries, then the multishard writer will read from exactly one shard.

We can't guarantee that we don't read across tablet boundaries yet, but eventually we will. So make_multishard_reader will call make_foreign_reader for tablets and this patch can then be reverted.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -5377,6 +5378,96 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+    return *leaving.begin();
+}
+
+inet_address storage_service::host2ip(locator::host_id host) {
```
> Could we avoid extending storage_service header by making this a static function, e.g.
static inet_address host2ip(const raft_address_map& map, locator::host_id host)?

If we fear extending storage_service, we should pimpl-ify it instead. This gives freedom to host2ip implementation in choosing dependencies and leaves usage sites clear of clutter.  

> have you considered reusing this?

I considered it, but it would introduce a preemption point which breaks atomicity of the code block, which is a problem. It also is unnecessary to loop here because read barrier ensures that local node knows about ips of all nodes in the topology due to the thing in topology_state_load(). And if it doesn't, streaming will be retried by the coordinator.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -5377,6 +5378,96 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+
+// Streams data from a given tablet replica to the pending tablet replica on this node.
+future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+    // The coordinator does not execute global token metadata barrier before jumping to "streaming" stage, so we need
```
It's better not to, because read barrier here is cheaper and more fault tolerant than the global barrier.

gleb-cloudius replied:
`raft_topology_cmd_handler` also has the same code. It also checks for outdated coordinator. Why no call `stream_tablet` through it instead of adding a separate RPC?

tgrabiec replied:
I considered it, but raft_topology_cmd_handler expects that commands are serialized, and here we want to stream in parallel.

gleb-cloudius replied:
Can you elaborate. As far as I see  `handle_tablet_streaming` sends a bunch rpcs to a bunch of nodes and waits for it to be complete. This is more or less what `raft_topology_cmd` does as well. In the case of `handle_tablet_streaming`  you have different params for different host, right? Each param says what should be streamed and where. Is this why separate RPC is needed? If yous then two question: can the node itself figure out this info like we do with regular streaming and if not then can we call `raft_topology_cmd` instead the new one but from iniside `handle_tablet_streaming`?

tgrabiec replied:
The RPC receiver figures out all the info except for which tablet to stream, and it must be the receiving replica. We want to stream tablets independently, so we have RPC per-tablet and per receiver rather than a single one. In this PR we stream tablets in tandem, but that will change in the next PR.

Regarding raft_topology_cmd reuse, it's answered here: https://github.com/scylladb/scylladb/pull/14601#discussion_r1274979394

tgrabiec commented on service/storage_service.cc:
```diff
@@ -5377,6 +5378,96 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+    return *ip;
+}
+
+// Streams data from a given tablet replica to the pending tablet replica on this node.
```
Yes, the comment can be confusing, I'll rephrase it.

tgrabiec replied:
> // Streams data to the pending tablet replica of a given tablet on this node.
> // The source tablet replica is determined from the current transition info of the tablet.
>

tgrabiec commented on locator/tablets.hh:
```diff
@@ -34,7 +34,29 @@ using token = dht::token;
-enum class tablet_id : size_t;
+struct tablet_id {
+    size_t id;
+    tablet_id() = default;
```
It seems unnecessary now, I'll drop it.

kbr-scylla commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+            auto src_host = candidate_nodes.back();
+            auto& src_node_info = nodes[src_host];
+
+            if (std::max(max_off_candidate_load, src_node_info.avg_load) == nodes[target].avg_load) {
```
Also please explain in a comment why this equality implies balance. I don't get it.

tgrabiec replied:
It is possible to have equality, if nodes have the same shard count.

I agree that the condition for balance can be improved, but it doesn't matter that much right now. In the worst case we will stop due to load inversion check. We could avoid building data structures if we had a smarter check, but this doesn't matter now when the load balancer is invoked rarely.

tgrabiec replied:
Which part is unclear? The goal is to equalize avg_load. max_off_candidate_load is higher than src_node_info.avg_load, so `std::max(max_off_candidate_load, src_node_info.avg_load)` is the max avg_load among non-target nodes. If that is equal to target's avg_load, it means all nodes have the same avg_load.

tgrabiec replied:
```
            // Check if all nodes reached the same avg_load. There are three sets of nodes: target, candidates (nodes_by_load) 
            // and off-candidates (removed from nodes_by_load). At any time, the avg_load for target is not greater than
            // that of any candidate, and avg_load of any candidate is not greater than that of any in the off-candidates set.
            // This is ensured by the fact that we remove candidates in the order of avg_load from the heap, and
            // because we prevent load inversion between candidate and target in the next check.
            // So the max avg_load of candidates is that of the current src_node_info, and max avg_load of off-candidates
            // is tracked in max_off_candidate_load. If max_off_candidate_load is equal to target's avg_load,
            // it means that all nodes have equal avg_load. We take the maximum with the current candidate in src_node_info
            // to handle the case of off-candidates being empty. In that case, max_off_candidate_load is 0.
```

avikivity replied:
> Is it even possible to achieve equality here in general? Especially since we're talking about doubles... We should at least use `diff < epsilon`.

It's not really important. We'll move to balancing on size (and later on CPU usage too, maybe).

I consider the algorithm here as a napkin sketch, what's more important is the code that drives the algorithm. Even that is just a sketch here as it doesn't work with edges.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+            // If there are 7 tablets and RF=3, each node must have 1 tablet replica.
+            // So node3 will have average load of 1, and node1 and node2 will have
+            // average shard load of 7.
+            lblogger.warn("Not possible to achieve balance.");
```
logging level is live-update-able via RESTful API. 

We do print information about node imbalance on info level, so you can at least see the impact.

I think tracing the root cause requires analysis by human, so he can enable more detailed logging or look at topology info in system tables.

avikivity replied:
It really shouldn't be a warning, perfect balance isn't a goal so we shouldn't worry about about not achieving it.

What we should have is a metric the quantifies imbalance, and then we can set alerts when it is large (and watch it jump when a node is added and then ~slowly~quickly drop again as tablets are streamed.

tgrabiec replied:
Ok, will change to info.

Adding metrics is on the todo list.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        /// valid across calls and only recalculating them when starting a new round with a new token metadata version.
+        ///
+        /// We can adjust the algorithm to work with multiple targets at a time. We should keep the candidate exclusion set
+        /// per-target.
```
No, there wasn't. Also, it's just a first version, not meant to be production-ready.

tgrabiec replied:
What I mean by that is that we care about correctness, but the algorithm doesn't have to be perfect in terms of decisions or performance. We know that it doesn't make perfect decisions and cannot achieve balance in some cases. For example, it doesn't balance within a node, or within a rack.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+                auto new_rack_load = rack_load[target_node->dc_rack().rack] + 1;
+                if (new_rack_load > max_rack_load) {
+                    lblogger.debug("candidate tablet {} skipped because it would increase load on rack {} to {}, max={}",
+                                   source_tablet, target_node->dc_rack().rack, new_rack_load, max_rack_load);
```
I think preserving fault-tolerance always takes precedence over balance. That's why we don't put two tablet replicas on the same node.

tgrabiec replied:
Yes, it can happen that rack constraint is preventing achieving balance.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+            auto& tmap = tmeta.get_tablet_map(source_tablet.table);
+            for (auto&& r : tmap.get_tablet_info(source_tablet.tablet).replicas) {
+                if (r.host == target) {
+                    lblogger.debug("candidate tablet {} skipped because it has a replica on target node", source_tablet, target);
```
Fixed by dropping target.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+            }
+
+            if (has_replica_on_target) {
+                lblogger.debug("candidate tablet {} skipped because it has a replica on target node", source_tablet);
```
Yes.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        }
+
+        future<> clear_gently() {
+            return utils::clear_gently(shards);
```
I don't think it's necessary, it's a vector of trivial types.

tgrabiec replied:
And with few elements (shards per node).

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        load_type avg_load = 0;
+
+        std::vector<shard_id> candidates; // heap which tracks most-loaded shards using candidates_cmp().
+        std::vector<shard_load> shards;
```
Yes.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+    using load_type = double;
+
+    struct shard_load {
+        size_t tablet_count;
```
I'll move the description.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+
+    struct shard_load {
+        size_t tablet_count;
+        std::unordered_set<global_tablet_id> candidates;
```
```
        // Tablets which still have a replica on this shard which are candidates for migrating away from this shard.
```

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        // The maximum per-shard load in terms of tablet count after the node would be internally balanced.
+        load_type avg_load = 0;
+
+        std::vector<shard_id> candidates; // heap which tracks most-loaded shards using candidates_cmp().
```
I'll rename to shards_by_load.

And candidate_nodes to nodes_by_load.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -922,6 +931,14 @@ class topology_coordinator {
         return {};
     }
 
+    inet_address id2ip(locator::host_id id) {
```
Given the logic in id2pi, and common expectation of throwing when mapping is not found, perhaps it should be a member of address_map, e.g. address_map::get()

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+            // Prevent load inversion which can lead to oscillations.
+            if (src_node_info.get_balanced_max_load(nodes[src_host].tablet_count - 1) <
+                    nodes[target].get_balanced_max_load(nodes[target].tablet_count + 1)) {
+                lblogger.debug("No more candidate nodes");
```
Ok

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        // which would be produced by the planner subsequently.
+
+        auto target_node = topo.find_node(target);
+        auto batch_size = target_node->get_shard_count();
```
Yes, I'm working on improving parallelizm right now.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        // heap which tracks most-loaded nodes in terms of avg_load.
+        std::vector<host_id> candidate_nodes;
+        candidate_nodes.reserve(nodes.size());
+        auto nodes_cmp = [&] (const host_id& a, const host_id& b) {
```
That's not exactly the same spirit, since the container is provided externally. candidates_cmp() lives at the same scope as the container. So in that spirit, nodes_cmp should live on the stack like right now, where the container is.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+
+        for (auto&& [table, tmap] : _tm->tablets().all_tables()) {
+            std::optional<tablet_id> tid = tmap.first_tablet();
+            if (!tmap.transitions().empty()) {
```
It's not a precondition. The precondition is that the plan is reflected, but doesn't have to be executed yet. When we implement concurrent streaming, load balancer will have to deal with active transitions.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+                        // This invariant is assumed later.
+                        if (replica.shard >= nodes[replica.host].shard_count) {
+                            auto gtid = global_tablet_id{table, *tid};
+                            throw std::runtime_error(format("Tablet {} replica {} targets non-existent shard", gtid, replica));
```
It can happen only due to a bug, but why on_internal_error()? We don't have to crash.

kbr-scylla replied:
But in production we won't crash, we'll throw. In tests we'll crash which is good for catching bugs early.

tgrabiec replied:
Ok, will change

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+
+        for (auto&& [table, tmap] : _tm->tablets().all_tables()) {
+            std::optional<tablet_id> tid = tmap.first_tablet();
+            for (const tablet_info& ti : tmap.tablets()) {
```
I work with the assumption that get_tablet_info() could get more expensive in the future.

but in any case we should have this iteration abstracted as for_each_tablet().

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        const locator::topology& topo = _tm->get_topology();
+        migration_plan plan;
+
+        // Prepare plans for each DC separately and combine them to be executed in parallel.
```
No, because DC's have a fixed RF. You cannot move a tablet away from a DC as that would reduce RF for that tablet.

kbr-scylla replied:
How will we handle schema changes that modify DC RFs?

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+};
+
+future<migration_plan> balance_tablets(token_metadata_ptr tm) {
+    load_balancer lb(tm);
```
This code is not performance-sensitive, so terseness wins.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        for (auto&& dc : topo.get_datacenters()) {
+            auto dc_plan = co_await make_plan(dc);
+            lblogger.info("Prepared {} migrations in DC {}", dc_plan.size(), dc);
+            std::copy(dc_plan.begin(), dc_plan.end(), std::back_inserter(plan));
```
ok

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -12,12 +12,419 @@
+        std::unordered_map<host_id, node_load> nodes;
+        topo.for_each_node([&] (const locator::node* node_ptr) {
+            if (node_ptr->get_state() == locator::node::state::normal && node_ptr->dc_rack().dc == dc) {
+                node_load& load = nodes[node_ptr->host_id()];
```
I'll take the liberty to ignore it, because here the fields are purposefully initialized in phases.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2597/

tgrabiec commented:
In v4:

 - Extracted documentation change to separate commit: [doc: Document tablet migration state machine and load balancer](https://github.com/scylladb/scylladb/pull/14601/commits/3c20fe79221289baeb6d6d6c065a27719b21b4d1)
 - Improved documentation in topology-over-raft.md with description of tablet migration state machine stage and its invariants.
 - Improved documentation of tablet_transition_stage in comments and commit log
 - tablet_allocator:
      - Moved algorithm description above class.
      - Added more comments describing algorithm and purpose of variables
      - Dropped redundant log message when skipping candidate due to replica on the target node
      - Renamed "candidates" to shard_by_load and nodes_by_load
      - Introduced tablet_map::for_each_tablet() and applied here
      - Rename get_balanced_max_load() -> get_avg_load()
      - Distinguish sibling "No more candidate nodes" logging with different message (one gets "load would be inverted" part). Also, log info about load averages of the two nodes involved. 
 - Decouple replica selection from stage in erm, for clarity:
      - Introduce write_replica_set_selector and read_replica_set_selector
      - Store selectors in tablet_transition_info, use them to select replicas in erm impl instead of stage
      - Convert stage to selectors when creating transition info 
 - Drop default constructor of tablet_id
 - Use std::move() instead of std::copy() on dc_plan
 - Fix typo in "eason"
 - Fix typo in "allocator" -> "allocate"
 - Use on_internal_error() in tablet_transition_stage_to_string()
 - Make tablet_stream_data RPC cancellable, cancel on topology coordinator abort
 - Fix comment on top of storage_service::stream_tablet() to not be confusing, hopefully now
 - Check for retry earlier in storage_service::stream_tablet()
 - Drop redundant commit which introduces release_guard()

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2598/

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -385,6 +388,12 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
                 return read_new_t::no;
             }
             switch (*state) {
+                case topology::transition_state::tablet_allow_write_both_read_old:
```
Don't all of these need `[[fallthrough]]`?

tgrabiec replied:
Right. The compilation didn't fail though.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2654/

tgrabiec commented:
In V5:

  - rebased
  - sprinkle more [[fallthrough]]
  - move read/write replica set selector calculation to tablet_transition_info ctor

gleb-cloudius commented on locator/tablets.cc:
```diff
@@ -267,7 +314,22 @@ class tablet_effective_replication_map : public effective_replication_map {
-        auto&& replicas = tablets.get_tablet_info(tablet).replicas;
+        auto* info = tablets.get_tablet_transition_info(tablet);
+        auto&& replicas = std::invoke([&] () -> const tablet_replica_set& {
+            if (!info) {
```
In which case there will be no `info`?

tgrabiec replied:
info is `tablet_transition_info`, which is only present during tablet migration. tablet_info is always present.

gleb-cloudius commented on locator/tablets.hh:
```diff
@@ -34,7 +34,28 @@ using token = dht::token;
 // Different tablets in subsequent token metadata version can have the same tablet_id.
 // When splitting a tablet, one of the new tablets (in the new token metadata version)
 // will have the same tablet_id as the old one.
-enum class tablet_id : size_t;
```
> The IDL compiler cannot deal with enum classes like this.

What is special about it? Can IDL compiler  be fixed?

avikivity commented on locator/tablets.cc:
```diff
@@ -135,6 +135,32 @@ const tablet_transition_info* tablet_map::get_tablet_transition_info(tablet_id i
     return &i->second;
```
@tgrabiec ^

avikivity replied:
I guess it's addressed.

tgrabiec replied:
It was addressed.

avikivity commented on locator/tablets.cc:
```diff
@@ -28,6 +28,53 @@ namespace locator {
+        case tablet_transition_stage::cleanup:
+            return read_replica_set_selector::next;
+    }
+    on_internal_error(tablet_logger, format("Invalid tablet transition stage", static_cast<int>(stage)));
```
Ditto

avikivity commented on locator/tablets.cc:
```diff
@@ -285,13 +347,40 @@ class tablet_effective_replication_map : public effective_replication_map {
+                return {get_endpoint_for_host_id(info->pending_replica.host)};
+            case write_replica_set_selector::next:
+                return {};
+        }
```
We should pre-calculate anything that changes only during edges, like info->writes.

tgrabiec replied:
You mean avoid the switch here? In this case, it saves only the switch. There's not much computation going on.

avikivity replied:
We could change it to return a span\<> later. But okay for now, since we'd need to copy into the return value anyway.

avikivity commented on locator/tablets.hh:
```diff
@@ -80,12 +80,24 @@ enum class tablet_transition_stage {
+    write_replica_set_selector writes;
+    read_replica_set_selector reads;
+
+    tablet_transition_info(tablet_transition_stage stage, tablet_replica_set next, tablet_replica pending_replica);
```
The publicness of everything weakens the safety here as the link between stage and selectors can be lost.

avikivity replied:
ok

avikivity commented on streaming/stream_reason.hh:
```diff
@@ -22,6 +22,7 @@ enum class stream_reason : uint8_t {
     rebuild,
     repair,
     replace,
+    tablet,
```
That's not a reason. We're migrating a tablet, not tableting anything.

tgrabiec replied:
Ok, will change to tablet_migration

avikivity commented on locator/tablets.hh:
```diff
@@ -216,6 +218,18 @@ public:
+    /// Calls a given function for each tablet in the map in token ownership order.
+    template <typename Func>
+    requires std::invocable<Func, tablet_id, const tablet_info&>
+    future<> for_each_tablet(Func func) const {
```
noncopyable_function? let's avoid bloat. Also lets us put this in .cc.

tgrabiec replied:
Ok

avikivity commented on locator/tablets.hh:
```diff
@@ -216,6 +218,18 @@ public:
+    future<> for_each_tablet(Func func) const {
+        std::optional<tablet_id> tid = first_tablet();
+        for (const tablet_info& ti : tablets()) {
+            co_await seastar::coroutine::maybe_yield();
```
Not needed as we have co_await below.

avikivity commented on locator/tablets.hh:
```diff
@@ -216,6 +218,18 @@ public:
+        for (const tablet_info& ti : tablets()) {
+            co_await seastar::coroutine::maybe_yield();
+            co_await futurize_invoke(func, *tid, ti);
+            tid = next_tablet(*tid);
```
tablet_map is immutable? Otherwise we need to know which lock needs to be held.

tgrabiec replied:
It is not immutable. When it's part of group0 state machine, it's locked by group0_guard.

avikivity replied:
ok

tgrabiec replied:
Well, no, it is actually immutable.

avikivity replied:
Even ok'er

avikivity commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -109,3 +109,40 @@ async def test_scans(manager: ManagerClient):
+    await manager.server_add()
+
+    await check()
+    time.sleep(5) # Give load balancer some time to do work
```
Let's loop here waiting for the migration to complete, peeking at the tablets table. We can keep check()ing meanwhile

tgrabiec replied:
It's not that simple. The fact that tablets are not in migration doesn't mean the load balancing is finished.

vladzcloudius commented:
@tgrabiec do you have some benchmark numbers? 
Specifically it would be interesting to see:
1) How much time it takes to add a new node up until all required data is streamed to it with tablets in:
   a) Idle cluster.
   b) Cluster under 50% of the maximum load.
   c) Cluster under 100% the maximum load.
2) How throughput/tail latency is impacted during the rebalancing in:
   a) Idle cluster.
   b) Cluster under 50% of the maximum load.
   c) Cluster under 100% the maximum load.

tgrabiec commented:
> @tgrabiec do you have some benchmark numbers?

No benchmarks yet. I plan to do them after posting more improvements. This version is known to be suboptimal in terms of performance.

tgrabiec commented:
In V6:
   - Added missing {} to format string
   - changed stream_reason::tablet to stream_reason::tablet_migration
   - demoted "Not possible to achieve balance" to info level
   - rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2683/

avikivity commented:
@benipeled promoter says `SUCCESS` but the integrated check report says sanity tests are pending.

avikivity commented:
Looks like a last-mile problem:

```
2:32:20  + curl -s -X POST -H 'Authorization: token ****' -H 'Accept: application/vnd.github.v3+json' -d '{"state": "success", "context": "Sanity Tests", "description": "Started by upstream project "scylla-master/scylla-ci" build number 2,683", "target_url": "https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/3226/"}' https://api.github.com/repos/scylladb/scylladb/statuses/ae8ffe23fc59cda3a29774e819837c6110b7c473
02:32:20  {
02:32:20    "message": "Problems parsing JSON",
02:32:20    "documentation_url": "https://docs.github.com/rest/commits/statuses#create-a-commit-status"
02:32:20  }
```

The json is indeed invalid, a string contains unescaped double quotes.

benipeled commented:
> @benipeled promoter says `SUCCESS` but the integrated check report says sanity tests are pending.

Solved on https://github.com/scylladb/scylla-pkg/pull/3499

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -5348,6 +5349,95 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+        throw std::runtime_error(format("No leaving replicas"));
+    }
+    if (leaving.size() > 1) {
+        throw std::runtime_error(format("More than one leaving replica"));
```
As far as I understand the code here checks exactly that: only one leaving replica for a tablet.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -5358,6 +5359,100 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+    // a barrier here to see the token metadata which is at least as recent as that of the sender.
+    auto& raft_server = _group0->group0_server();
+    co_await raft_server.read_barrier(&_abort_source);
+
```
Shouldn't you check that the coordinator is not outdated?

tgrabiec replied:
Instead of that, I check whether the request matches what's in local tablet transition info. Here's the line of thought. Instead of an RPC, we could trigger streaming from the local change of tablet_transition_info, as all the decision follows from that. There is no need for global token metadata barrier in the "streaming" stage, because it's performed in the previous stage. So it doesn't matter if the coordinator is still the same, as long as we execute what is in the current tablet_transition_info. If that migration is aborted (not impl yet) the coordinator would execute global token metadata barrier to ensure all replicas notice the abort stage and old RPCs don't run.

tgrabiec replied:
> So why pass it at all? Just take it from there and check for outdated RPC the same way topology_cmd does?

You mean why pass tablet id instead of starting all sessions which target the RPC-handling node? Even though in this PR we stream all tablets in tandem, later we will wan to stream independently, so the coordinator needs to know when a given tablet finished migrating, not just all tablets on a given replica.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1149,6 +1172,133 @@ class topology_coordinator {
+
+    // The state "streaming" is needed to ensure that stale stream_tablet() RPC doesn't
+    // get admitted before global_tablet_token_metadata_barrier() is finished in
+    // handle_tablet_write_both_read_old().
```
Can you elaborate. I am not sure I understand how having separate "streaming" stage guards from stale RPCs. Wouldn't sending streaming command through `raft_topology_cmd` rpc solve this without the state?

tgrabiec replied:
I cannot reuse current raft_topology_cmd because it doesn't support parallelism

gleb-cloudius replied:
> Having "streaming" stage means there is no work for the coordinator other than triggering streaming, so as soon as the relica sees "streaming" stage it means it's ok to trigger streaming and we don't need to worry about guarding against old rpcs.

What prevents very old RPC to arrive when replica is in the "streaming" stage. I guess I am missing something here.

gleb-cloudius replied:
> I cannot reuse current raft_topology_cmd because it doesn't support parallelism

In what way it does not support it? You cannot call `raft_topology_cmd` multiple times on the same node simultaneously by different RPC (because of `_raft_topology_cmd_handler_state presumably)? But is it really needed? Can we pass all the work to a node in one RPC? Will be more efficient as well.

tgrabiec replied:
> What prevents very old RPC to arrive when replica is in the "streaming" stage. I guess I am missing something here.

Nothing prevents that, but in this implementation it's ok if the old RPC does trigger streaming, because it will do what the coordinator wants to do right now anyway. The new RPC will block on that streaming (retry path).

tgrabiec replied:
> In what way it does not support it? You cannot call `raft_topology_cmd` multiple times on the same node simultaneously by different RPC (because of `_raft_topology_cmd_handler_state presumably)? But is it really needed? Can we pass all the work to a node in one RPC? Will be more efficient as well.

Yes, we can have only one raft_topology_cmd per node at a time due to the check for cmd_index. If we submitted them in parallel, one of them could get reordered and get rejected. We could batch requests per node in this PR, because we stream in tandem, but this will change in the next PR, we want to stream tablets independently, and may start new sessions while old sessions are still running.

gleb-cloudius replied:
> Nothing prevents that, but in this implementation it's ok if the old RPC does trigger streaming, because it will do what the coordinator wants to do right now anyway. The new RPC will block on that streaming (retry path).

So why do you prefer to have additional "streaming" state instead of filtering out outdated RPCs?

gleb-cloudius replied:
> Yes, we can have only one raft_topology_cmd per node at a time due to the check for cmd_index. If we submitted them in parallel, one of them could get reordered and get rejected. We could batch requests per node in this PR, because we stream in tandem, but this will change in the next PR, we want to stream tablets independently, and may start new sessions while old sessions are still running.

Would adding a lock around cmd_index help?

tgrabiec replied:
> So why do you prefer to have additional "streaming" state instead of filtering out outdated RPCs?

To avoid maintaining extra per-tablet state on each replica. Adding a stage seems simpler.

gleb-cloudius replied:
> > So why do you prefer to have additional "streaming" state instead of filtering out outdated RPCs?
> 
> To avoid maintaining extra per-tablet state on each replica. Adding a stage seems simpler.

Why should it be per-tablet as opposite to global like now?

tgrabiec replied:
> Would adding a lock around cmd_index help?

I don't immediately see how. Suppose you start streaming for tablet X, and immediately after but in in the same batch, start streaming for Y. RPCs for X and Y will race on the receiving replica. They may get reordered while executing the read barrier. Unless we somehow prevent reordering in all the layers involved, but that seems fragile.

tgrabiec replied:
> Why should it be per-tablet as opposite to global like now?

Because of the same problem with parallelism and reordering we talk about regarding raft_topology_cmd.

gleb-cloudius replied:
> I don't immediately see how. Suppose you start streaming for tablet X, and immediately after but not in the same batch, start streaming for Y. RPCs for X and Y will race on the receiving replica. They may get reordered while executing the read barrier. Unless we somehow prevent reordering in all the layers involved, but that seems fragile.

It can be solved by a lock on rpc handler level. On RPC level itself reordering may only happen in case of a failed connection which is rare and can be handled by simply retrying an rpc on the coordinator. But I see what you mean. It is unfortunate though that outdated RPCs can be executed. Much harder to reason about things.
